Add ca_generate() function

ddt-tdd opened the pull request:
Hi,

I added the support to generate a certificate from a CA certificate. It would be kind to merge this.
This allows to import the Root CA in browser so the rust library is able to issue SSL certificates which are trusted in the browser.

Greetings,
DDT

sfackler commented on openssl/src/x509/mod.rs:
```diff
@@ -335,7 +335,7 @@ impl X509Generator {
 
         unsafe {
-            let x509 = ffi::X509_new();
+            let x509 = ffi::X509_new();            
```
Trailing whitespace

sfackler commented on openssl/src/x509/mod.rs:
```diff
@@ -390,6 +390,69 @@ impl X509Generator {
         }
     }
 
+    /// Sets the certificate public-key, then self-sign and return it
```
It seems like a good bit of the code between this and sign (and maybe others?) can be shared.

ddt-tdd replied:
Hi Steven,

I refactored the code to prevent DRY-pattern.

I removed the ca_generate() function and added a set_ca(). I extended the
struct of X509Generate with ca: Option<(&'ctx X509<'ctx>, &'ctx PKey)>
I think it is cleaner from usage point of view, but I had to add the
lifetime specifier of X509 to the X509Generate struct and all function who
return the X509Generate object.
I asked a new pull request. Any comments are welcome.

Greetings,
DDT

2016-06-19 2:02 GMT+02:00 Steven Fackler <notifications@github.com>:

> In openssl/src/x509/mod.rs
> <https://github.com/sfackler/rust-openssl/pull/411#discussion_r67608303>:
>
> > @@ -390,6 +390,69 @@ impl X509Generator {
> >          }
> >      }
> >
> > +    /// Sets the certificate public-key, then self-sign and return it
>
> It seems like a good bit of the code between this and sign (and maybe
> others?) can be shared.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sfackler/rust-openssl/pull/411/files/338a02cb2e01401368260ceee3a0a3f2735d3f2a#r67608303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABEe89Rnrz_qJ2L86NZhekpd62bYXQ23ks5qNIcbgaJpZM4I20vR>
> .
>

ddt-tdd commented:
Hi Steven,

I refactored the code to prevent DRY-pattern.

I removed the ca_generate() function and added a set_ca(). I extended the struct of X509Generate with ca: Option<(&'ctx X509<'ctx>, &'ctx PKey)>
I think it is cleaner from usage point of view, but I had to add the lifetime specifier of X509 to the X509Generate struct and all function who return the X509Generate object.
I asked a new pull request. Any comments are welcome.

Greetings,
DDT

sfackler commented:
Sorry for missing this!

This needs a rebase, but then looks good to go!

ddt-tdd commented:
That's OK. But I see that some Travis builds are failing. I checked the logs on travis and it seems not really linked to modifications I made.
Do you want me to perform the rebasing?

sfackler commented:
Yep, could you rebase?

ddt-tdd commented:
I've perform the rebase of the code.

sfackler commented on openssl/src/x509/mod.rs:
````diff
@@ -129,32 +129,34 @@ impl X509StoreContext {
 /// let pkey_pem = pkey.private_key_to_pem().unwrap();
 /// ```
-pub struct X509Generator {
+pub struct X509Generator<'ctx> {
````
Adding this type parameter is unfortunately a breaking change. Maybe we should have `set_ca` take the cert and pkey by value and fix it when we cut the next breaking release?

ddt-tdd replied:
That's OK. If I understand well, the lifetime specifier <'ctx> breaks the
some implementation. All previous tests do work without any modification.
You changed the interface and removed the generate() function, so
everything is done through the sign function. Maybe I can use the sign
function of the CA to sign my SSL certificate and refactor the code like
that.
I'll check that.

2016-08-14 18:31 GMT+02:00 Steven Fackler <notifications@github.com>:

> In openssl/src/x509/mod.rs
> <https://github.com/sfackler/rust-openssl/pull/411#discussion_r74704755>:
>
> > @@ -129,32 +129,34 @@ impl X509StoreContext {
> >  /// let cert_pem = cert.to_pem().unwrap();
> >  /// let pkey_pem = pkey.private_key_to_pem().unwrap();
> >  /// ```
> > -pub struct X509Generator {
> > +pub struct X509Generator<'ctx> {
>
> Adding this type parameter is unfortunately a breaking change. Maybe we
> should have set_ca take the cert and pkey by value and fix it when we cut
> the next breaking release?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/sfackler/rust-openssl/pull/411/files/8d9273ba3de769dc580363447d0dd4fdf14329e7#r74704755>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABEe8_NlppjoTX3LXchn_gVA1vemeMC7ks5qf0LNgaJpZM4I20vR>
> .
>

sfackler replied:
Another possibility could be to add a separate `ca_sign` method that takes in the CA components if transferring ownership of the CA to the generator won't work.

ddt-tdd commented:
Hi,

I removed the lifetime parameter and created a sign_by_ca function, with a test.
Maybe a bad question here, but will hyper http server migrate to 0.8.x of openssl?

Greetings,
DDT

sfackler commented:
Hopefully - that seems like a question for @seanmonstar though

sfackler commented on openssl/src/x509/mod.rs:
```diff
@@ -349,6 +350,61 @@ impl X509Generator {
 
+    /// Sets the certificate public-key, then signs it by the CA key and certificate and return it
+    /// Note: That the bit-length of the private key is used (set_bitlength is ignored)
+    pub fn sign_by_ca(&self, subject_pkey: &PKey, ca_cert: &X509, ca_pkey: &PKey) -> Result<X509, ErrorStack> {
```
Would it be possible to reduce the code duplication in the internals of `sign` and `sign_by_ca`?

ddt-tdd commented:
Hi Steven,

I've refactored using the DRY principle as much as possible. I also rebased to the 0.8.3. If you've time, can you merge the sign_by_ca functionality.

Greetings,
DDT

sfackler commented:
Oh hey, sorry for letting this sit for so long! I'm actually planning on deprecating all of this infrastructure in favor of a more direct interface to the underlying APIs. Can you check if https://github.com/sfackler/rust-openssl/pull/515 has all of the functionality you need?

sfackler commented:
I've landed #515 - please file an issue if anything's missing with what it provides!
